pkg/controller/machineset/machineset_controller_test.go

```diff
@@ -17,196 +17,90 @@ limitations under the License.
 
-				// TODO (robertbailey): Figure out why the control loop isn't working as expected.
-				/*
-					g.Eventually(func() int {
-						if err := c.List(context.TODO(), machines); err != nil {
-							return -1
-						}
-						return len(machines.Items)
-					}, timeout).Should(gomega.BeEquivalentTo(replicas))
-				*/
-			},
-		},
-	}
-
-	for _, tc := range testCases {
-		t.Logf("Running %q testcase", tc.name)
-		func() {
-			if err := c.Create(context.TODO(), tc.instance); err != nil {
-				t.Errorf("error creating instance: %v", err)
-			}
+var _ = Describe("MachineSet Reconciler", func() {
+	namespace := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "ms-test"}}
 
-			defer func() {
-				err := c.Delete(context.TODO(), tc.instance)
-				if err != nil {
-					t.Logf("error deleting %v: %v", tc.instance, err)
-				}
+	BeforeEach(func() {
+		By("Creating the namespace")
+		Expect(k8sClient.Create(ctx, namespace)).To(Succeed())
+	})
 
-				select {
-				case recv := <-requests:
-					if recv != tc.expectedRequest {
-						t.Error("received request does not match expected request")
-					}
+	AfterEach(func() {
+		By("Deleting the namespace")
+		Expect(k8sClient.Delete(ctx, namespace)).To(Succeed())
+	})
 
-					t.Logf("Last reconcile %v", <-requests)
-				case <-time.After(timeout):
-					t.Error("timed out waiting for request")
-				}
-			}()
+	It("Should reconcile a MachineSet", func() {
+		replicas := int32(2)
+		labels := map[string]string{"foo": "bar"}
 
-			select {
-			case recv := <-requests:
-				if recv != tc.expectedRequest {
-					t.Errorf("received request does not match expected request. Got: %v, expected: %v", recv, tc.expectedRequest)
-				}
-			case <-time.After(timeout):
-				t.Error("timed out waiting for request")
-			}
+		instance := &machinev1.MachineSet{
+			ObjectMeta: metav1.ObjectMeta{Name: "foo", Namespace: namespace.Name},
+			Spec: machinev1.MachineSetSpec{
+				Replicas: &replicas,
+				Selector: metav1.LabelSelector{
+					MatchLabels: labels,
+				},
+				Template: machinev1.MachineTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: labels,
+					},
+				},
+			},
+		}
 
-			tc.verifyFnc()
+		By("Creating the MachineSet")
+		Expect(k8sClient.Create(ctx, instance)).To(Succeed())
+		defer func() {
+			Expect(k8sClient.Delete(ctx, instance)).To(Succeed())
 		}()
-	}
-}
 
-func expectInt(t *testing.T, expect int, fn func(context.Context) int) {
-	t.Helper()
+		machines := &machinev1.MachineList{}
 
-	ctx, cancel := context.WithTimeout(context.TODO(), timeout)
-	defer cancel()
-
-	intCh := make(chan int)
-
-	go func() { intCh <- fn(ctx) }()
+		By("Verifying that we have 2 replicas")
+		Eventually(func() int {
+			if err := k8sClient.List(ctx, machines, client.InNamespace(namespace.Name)); err != nil {
+				return -1
+			}
+			return len(machines.Items)
+		}, timeout).Should(BeEquivalentTo(replicas))
```

Comment:
You can return a second parameter here (I learnt this recently) and Eventually will check the non-first params are their zero values/nil, that way you don't have to make up "error" values as we have done before. Might be worth doing the same on the others so that errors are surfaced properly
```suggestion
		Eventually(func() (int, error) {
			if err := k8sClient.List(ctx, machines, client.InNamespace(namespace.Name)); err != nil {
				return 0, err
			}
			return len(machines.Items), nil
		}, timeout).Should(BeEquivalentTo(replicas))
```

Reply:
👍